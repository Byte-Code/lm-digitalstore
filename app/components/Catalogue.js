import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { Link } from 'react-router';
import { Map, List } from 'immutable';
import glamorous from 'glamorous';

import ProductBadge from './ProductBadge';
import SellingAidsBadge from './SellingAidsBadge';
import FilterBar from './FilterBar';

export default class Catalogue extends Component {
  static propTypes = {
    params: PropTypes.shape({ categoryCode: PropTypes.string.isRequired }).isRequired,
    requestFetchCategory: PropTypes.func.isRequired,
    clearProductList: PropTypes.func.isRequired,
    categoryInfo: ImmutablePropTypes.map,
    products: ImmutablePropTypes.list,
    filterMap: ImmutablePropTypes.map.isRequired,
    toggleAid: PropTypes.func.isRequired,
    toggleFilter: PropTypes.func.isRequired,
    toggleAvailability: PropTypes.func.isRequired,
    resetFilters: PropTypes.func.isRequired,
    initFilters: PropTypes.func.isRequired,
    toggleFiltersDialog: PropTypes.func.isRequired,
    resetTempFilters: PropTypes.func.isRequired,
    isDialogOpen: PropTypes.bool.isRequired
  };

  static defaultProps = {
    categoryInfo: Map(),
    products: List()
  };

  componentDidMount() {
    const { params: { categoryCode }, requestFetchCategory, initFilters } = this.props;
    initFilters();
    requestFetchCategory(categoryCode);
  }

  componentWillReceiveProps(nextProps) {
    const { params: { categoryCode }, requestFetchCategory, initFilters } = nextProps;
    if (categoryCode !== this.props.params.categoryCode) {
      initFilters();
      requestFetchCategory(categoryCode);
    }
  }

  componentWillUnmount() {
    this.props.clearProductList();
  }

  renderProducts() {
    const { products } = this.props;
    return products.map(p => (
      <Link to={`product/${p.get('code')}`} key={p.get('code')}>
        <ProductBadge productInfo={p} />
      </Link>
    ));
  }

  render() {
    const {
      categoryInfo,
      filterMap,
      toggleAid,
      resetFilters,
      toggleAvailability,
      toggleFilter,
      toggleFiltersDialog,
      isDialogOpen,
      resetTempFilters
    } = this.props;

    if (categoryInfo.isEmpty()) {
      return null;
    }

    const catName = categoryInfo.get('name');
    const sellingAids = categoryInfo.getIn(['sellingAidsProducts', 0]) || Map();
    const facetFilters = categoryInfo.get('facetFilters') || List();
    const filterGroups = facetFilters.filterNot(g => g.get('group') === 'Prezzo');
    const activeAid = filterMap.get('aid');

    return (
      <div>
        <Header>
          <h1>{catName}</h1>
        </Header>
        <SellingAidsBadge sellingAids={sellingAids} onToggle={toggleAid} activeAid={activeAid} />
        <FilterBar
          filterGroups={filterGroups}
          resetFilters={resetFilters}
          filterMap={filterMap}
          toggleFilter={toggleFilter}
          toggleAvailability={toggleAvailability}
          toggleFiltersDialog={toggleFiltersDialog}
          isDialogOpen={isDialogOpen}
          resetTempFilters={resetTempFilters}
        />
        <ProductSlider opacity={isDialogOpen}>
          <FakeMarginDiv />
          {this.renderProducts()}
          <FakeMarginDiv />
        </ProductSlider>
      </div>
    );
  }
}

const FakeMarginDiv = glamorous.div({
  height: '100%',
  width: '40px'
});

const Header = glamorous.div({
  width: '100%',
  height: '184px',
  background: '#f7f7f7',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  '&>h1': {
    fontSize: '48px'
  }
});

const ProductSlider = glamorous.div(({ opacity = false }) => ({
  marginTop: '5%',
  display: 'flex',
  overflowX: 'auto',
  flexFlow: 'column wrap',
  height: '1246px',
  opacity: opacity ? 0.17 : 1,
  '&>a': {
    width: '405px',
    height: '593px',
    marginRight: '20px',
    '&:nth-child(odd)': {
      marginBottom: '60px'
    }
  }
}));
