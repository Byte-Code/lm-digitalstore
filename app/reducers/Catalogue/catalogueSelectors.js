export default function productList(state) {
  return state;
}
