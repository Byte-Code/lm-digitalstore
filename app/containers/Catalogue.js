import { connect } from 'react-redux';

import Catalogue from '../components/Catalogue';
import { requestFetchCategory } from '../actions/categoryActions';
import { requestFetchProducts, clearProductList } from '../actions/productListActions';
import * as filtersActions from '../actions/filtersActions';
import { getCatalogueProducts, getFilterMap, getCategory, getDialogStatus } from '../reducers/selectors';
import { setAnalyticsProductClick } from '../actions/analyticsActions';

const mapStateToProps = (state, ownProps) => {
  const { params: { categoryCode } } = ownProps;
  return {
    categoryInfo: getCategory(state, categoryCode),
    products: getCatalogueProducts()(state, categoryCode),
    filterMap: getFilterMap(state),
    isDialogOpen: getDialogStatus(state)
  };
};

const mapDispatchToProps = {
  requestFetchCategory,
  requestFetchProducts,
  clearProductList,
  ...filtersActions,
  setAnalyticsProductClick
};

export default connect(mapStateToProps, mapDispatchToProps)(Catalogue);
