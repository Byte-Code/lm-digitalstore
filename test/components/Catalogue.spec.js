import React from 'react';
import { fromJS, Map, List } from 'immutable';
import { shallow, mount } from 'enzyme';

import Catalogue from '../../app/components/Catalogue';

const params = { categoryCode: 'CAT123' };
const requestFetchCategory = jest.fn();
const initFilters = jest.fn();
const products = fromJS([
  { code: 1, mainImage: 'image1', name: 'product1', marketingAttributes: {} },
  { code: 2, mainImage: 'image2', name: 'product2', marketingAttributes: {} },
  { code: 3, mainImage: 'image3', name: 'product3', marketingAttributes: {} }
]);
const categoryInfo = fromJS({
  name: 'cat123',
  sellingAidsProducts: List(),
  facetFilters: List()
});
const router = { location: {} };
const filterMap = Map();

jest.mock('../../app/analytics/AnalyticsService');
jest.mock('../../app/CommandLineOptions', () => ({
  isDebugMode: jest.fn()
}));

describe('catalogue', () => {
  it('should render null when categoryInfo is empty or undefined', () => {
    const result = shallow(
      <Catalogue
        params={params}
        filterMap={filterMap}
        products={products}
        requestFetchCategory={requestFetchCategory}
        router={router}
        toggleAid={() => {}}
        toggleFilter={() => {}}
        toggleAvailability={() => {}}
        resetFilters={() => {}}
        initFilters={() => {}}
        applyFilters={() => {}}
        clearProductList={() => {}}
        toggleFiltersDialog={() => {}}
        isDialogOpen
      />
    );
    expect(result).toMatchSnapshot();
  });

  it('should render properly when categoryInfo is not empty', () => {
    const result = shallow(
      <Catalogue
        params={params}
        categoryInfo={categoryInfo}
        filterMap={filterMap}
        products={products}
        requestFetchCategory={requestFetchCategory}
        router={router}
        toggleAid={() => {}}
        toggleFilter={() => {}}
        toggleAvailability={() => {}}
        resetFilters={() => {}}
        initFilters={() => {}}
        applyFilters={() => {}}
        clearProductList={() => {}}
        toggleFiltersDialog={() => {}}
        isDialogOpen
      />
    );
    expect(result).toMatchSnapshot();
  });

  it('should call requestFetchCategory after mount', () => {
    mount(
      <Catalogue
        params={params}
        categoryInfo={categoryInfo}
        filterMap={filterMap}
        products={products}
        requestFetchCategory={requestFetchCategory}
        router={router}
        initFilters={initFilters}
        toggleAid={() => {}}
        toggleFilter={() => {}}
        toggleAvailability={() => {}}
        resetFilters={() => {}}
        applyFilters={() => {}}
        clearProductList={() => {}}
        toggleFiltersDialog={() => {}}
        isDialogOpen
      />
    );
    expect(initFilters).toHaveBeenCalled();
    expect(requestFetchCategory).toHaveBeenCalledWith('CAT123');
  });

  it('should call initFilters and requestFetchCategory when categoryCode changes', () => {
    const result = mount(
      <Catalogue
        params={params}
        categoryInfo={categoryInfo}
        filterMap={filterMap}
        products={products}
        router={router}
        requestFetchCategory={requestFetchCategory}
        initFilters={initFilters}
        toggleAid={() => {}}
        toggleFilter={() => {}}
        toggleAvailability={() => {}}
        resetFilters={() => {}}
        applyFilters={() => {}}
        clearProductList={() => {}}
        isDialogOpen
        toggleFiltersDialog={() => {}}
      />
    );
    const nextProps = {
      params: {
        categoryCode: 'CAT456'
      },
      products: List(),
      filterMap: Map({ filters: List() })
    };
    result.setProps(nextProps);
    expect(initFilters).toHaveBeenCalled();
    expect(requestFetchCategory).toHaveBeenCalledWith('CAT456');
  });
});
