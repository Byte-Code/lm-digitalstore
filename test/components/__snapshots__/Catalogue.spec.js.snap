// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`catalogue should render null when categoryInfo is empty or undefined 1`] = `null`;

exports[`catalogue should render properly when categoryInfo is not empty 1`] = `
<div>
  <glamorous(div)>
    <h1>
      cat123
    </h1>
  </glamorous(div)>
  <SellingAidsBadge
    activeAid=""
    onToggle={[Function]}
    sellingAids={Object {}}
  />
  <FilterBar
    filterGroups={Array []}
    filterMap={Object {}}
    isDialogOpen={true}
    resetFilters={[Function]}
    toggleAvailability={[Function]}
    toggleFilter={[Function]}
    toggleFiltersDialog={[Function]}
  />
  <Swipeable
    delta={50}
    disabled={false}
    flickThreshold={0.6}
    nodeName="div"
    onSwipingLeft={[Function]}
    onSwipingRight={[Function]}
    preventDefaultTouchmoveEvent={true}
    stopPropagation={false}
  >
    <glamorous(div)
      opacity={true}
    />
  </Swipeable>
</div>
`;
